ddtrace/internal/processor/trace.py

```diff
@@ -226,3 +227,34 @@ def shutdown(self, timeout):
         except ServiceStatusError:
             # It's possible the writer never got started in the first place :(
             pass
+
+
+@attr.s
+class SingleSpanSamplingProcessor(SpanProcessor):
+    """SpanProcessor for sampling single spans:
+    - Users can specify span rules with service, name, sample rate, and rate limit.
+    - Those rules are used to assess all spans that are not sampled by trace sampling.
+    If the rules assess that a span should be kept, we add specific tags to it.
+    - The single spans that will be kept are sent in separate payloads if stats computation is enabled.
+    If stats computation is not enabled, then the single spans will be sent along with normal trace payloads.
+    """
+
+    rules = attr.ib(type=(List[SpanSamplingRule]))
+
+    def on_span_start(self, span):
+        # type: (Span) -> None
+        pass
+
+    def on_span_finish(self, span):
+        # type: (Span) -> None
+        # only sample if the span isn't already going to be sampled by trace sampler
+        if span.context.sampling_priority is not None and span.context.sampling_priority <= 0:
+            for rule in self.rules:
+                rule.sample(span)
+                # If we matched a rule, then don't try to apply any further rules
+                if rule.match(span):
+                    break
```

Comment:
is this backwards?

what if no rules match? does that mean we still end up applying a rule?

I would have expected something more like:

```suggestion
                if rule.match(span):
                    rule.sample(span)
                    # If we matched a rule, then don't try to apply any further rules
                    break               
```

Reply:
In `rule.sample` we actually call `rule.match`, so if it doesn't match, we just don't do anything to the span. However maybe I should refactor so that we don't call rule.match in rule.sample, and then just use your code because it helps us keep from trying to match twice.... going with that. Thanks for the call out, this felt weird when doing it but I couldn't pinpoint the solution then.

Reply:
oh, that is confusing that we call `match` in sample